jbmc/src/java_bytecode/java_types.cpp

```diff
@@ -182,6 +182,66 @@ bool is_multidim_java_array_type(const typet &type)
                                        to_struct_tag_type(type.subtype())));
 }
 
+static bool is_reference_array_type(const typet &type)
+{
+  const auto *reference_type =
+    type_try_dynamic_cast<java_reference_typet>(type);
+  if(!reference_type)
+    return false;
+  const auto *struct_tag_type =
+    type_try_dynamic_cast<struct_tag_typet>(reference_type->subtype());
+  return struct_tag_type &&
+         struct_tag_type->get_identifier() == JAVA_REFERENCE_ARRAY_CLASSID;
+}
+
+static typet get_array_element_subtype(const typet &type)
+{
+  const typet &array_type = to_java_reference_type(type).subtype();
+  return java_array_element_type(to_struct_tag_type(array_type));
+}
+
+/// Returns the underlying element type and array dimensionality of Java struct
+/// \p type. If this is not an array type the return value will be `{type, 0}`.
+std::pair<typet, std::size_t>
+java_array_dimension_and_element_type(const struct_tag_typet &type)
```

Comment:
:no_entry_sign: out of band discussion - this function has changed name, but still only returns non-zero dimension for *reference* type arrays. Either the name should remain `java_reference_array_dimension...` or it should handle primitive array dimensions